### DISCUSSION TITLE
.CLS and .BAS file support

### DISCUSSION THREAD

**Discussion by WaynePhillipsEA** [twinbasic/twinbasic#257] [General]
As per the title, v0.10.3377 now includes preliminary support for .CLS and .BAS files.

Be aware that the files are currently interpreted with UTF-8 encoding, which is not correct.  As these are legacy formats, tB will (eventually) be treating them exactly as VB6/VBA does, which is as ANSI files in the current system codepage.  As such, unicode will not be supported at all in these legacy formats.  A good reason to upgrade to the .twin format, I say :)

You can drag & drop a .bas/cls file into tB now, or if you specify .bas or cls in the 'add a component' dialog, it will create a legacy file for you...

<img width="220" alt="tb_CLS_0" src="https://user-images.githubusercontent.com/25738109/124295543-ddcd0680-db50-11eb-86ab-262265c18956.png">

<img width="390" alt="tb_CLS_1" src="https://user-images.githubusercontent.com/25738109/124295623-f76e4e00-db50-11eb-916d-13ba7b8f1141.png">

**Comment by Kr00l** [comment]
The tab indentation is only the half in a new legacy .CLS compared to a .twin or a .BAS file ?
So since .CLS is not same as in .BAS I suspect it's a bug ?

**Reply by WaynePhillipsEA** [reply]
I think VS Code is trying to be clever there, matching the BEGIN block indentation.  I will check tomorrow to see what I can do about that

**Reply by WaynePhillipsEA** [reply]
Fixed in v0.10.3412.  I changed the spaces in the BEGIN header block to tabs so that VS code doesn't do auto-detection of the 2-space indentation for the document.  VB6 allows for the tabs.

**Comment by FullValueRider** [comment]
Drag and drop doesn't work for me.  Neither can I find an add component command in the command palette.

**Reply by FullValueRider** [reply]
OK so having posted the above I immediately found the add component dialog <sigh>

**Reply by WaynePhillipsEA** [reply]
Drag and drop should be working:

![tb_dragDropCls](https://user-images.githubusercontent.com/25738109/124355793-ea159a00-dc0a-11eb-833e-0debd6e8191c.gif)

**Reply by FullValueRider** [reply]
I tried following the example above but it does not work for me. Instead off seeing the copy annotation I just get a circle with a bar across it which changes to copy once I move off the VScode window.  At the moment the icon associated with the dragged file is that of Notepad++.

**Reply by FullValueRider** [reply]
I think I found the problem.  If I create a new twinbasic project from within VS Code I cannot drag and drop.  If I save the project/close VS Code, then reopen the project , drag and drop works fine.

**Reply by WaynePhillipsEA** [reply]
Works both ways for me.  When you start a new twinbasic project, I don't suppose you've started VS Code with the 'Run as administrator' Windows option?  That would stop the drag & drop from working because the explorer.exe cannot interact with the higher privileged VS Code process.

**Reply by FullValueRider** [reply]
Bingo.  That's the explanation.  I automatically use Administrator mode.  A habit I acquired from Visual Studio where c# code for a library would not compile unless you were in administrator mode.

**Comment by WaynePhillipsEA** [comment]
> Be aware that the files are currently interpreted with UTF-8 encoding, which is not correct. As these are legacy formats, tB will (eventually) be treating them exactly as VB6/VBA does, which is as ANSI files in the current system codepage. As such, unicode will not be supported at all in these legacy formats. A good reason to upgrade to the .twin format, I say :)

This is now fixed in v0.10.4503.

When editing BAS/CLS files in VS Code, files are massaged to UTF8 encoding, but when persisted to (or loaded from) storage they are ANSI-encoded using the system codepage, which provides backwards compatibility with VB6/VBA.

If you require Unicode support in a source file, you must use the .twin format, since UTF8 is fully supported there.

**Reply by wqweto** [reply]
Does this allow importing (via drag&drop) BAS/CLS files which contain string literals with symbols in high ANSI codepage (like cyrillic above &H80 character codes) now?

Once such ANSI files are imported (using local system locale mentioned in #353) what was the rationale for making the extra effort to **not** allow Unicode content getting persisted in the VFS for these particular file extensions (requiring manually renaming them to .twin) provided that ANSI codepage vs UTF-8 incompatibility with VBx was resolved by fixing the import?

Was it the idea to allow round-tripping BAS/CLS sources from TB projects back to VBx ones without loss of encoding?

**Reply by WaynePhillipsEA** [reply]
Yes, drag & drop should work fine for that now.

>Was it the idea to allow round-tripping BAS/CLS sources from TB projects back to VBx ones without loss of encoding?

Yes, pretty much.  There were also some technical reasons;

The VS Code FileSystemProvider API exposes only a simple writeFile event for when files get updated/created, and so we cannot accurately determine if a file was dropped into the file system (which would be ANSI based), or created via some other means through the VS Code UI (which _might_ be UTF8 based depending on where it came from).  So the decision was made to not try to upgrade the ANSI files to UTF8 on **_import_**, as we cannot accurately determine that the conversion is necessary at that point.  Instead we persist the imported files unaltered (as ANSI).  The ANSI files then get converted 'live' to UTF8 for VS Code consumption as and when when the documents are requested, and then stored back to ANSI on save from the VS Code UI.  If characters outside of the system codepage are used in a BAS/CLS file, the save will fail and a warning is issued.

We did consider appending the UTF8 BOM on saving to allow us to use and identify UTF8 BAS/CLS files, but it is the round-tripping compatibility with VBx that put us off that approach particularly in light of #62 where we don't want to 'upgrade' the BAS/CLS files on disk to UTF8 if they are still used in a VB6 project.

**Reply by DavidHollman** [reply]
I've brought up version control "compatibility" before and this approach is another + for that. Automatically re-saving the files in a different format would suddenly cause all sorts of noise there.
